realsense/plugin.cpp

```diff
@@ -0,0 +1,179 @@
+						last_iteration_accel = accel_type_.iteration;
+						rs2_vector accel = *accel_type_.accel_data;
+						rs2::motion_frame gyro = mf;
+						ts = gyro.get_timestamp();
+						gyro_data = gyro.get_motion_data();
+
+						// IMU data
+						la = {accel.x, accel.y, accel.z};
+						av = {gyro_data.x, gyro_data.y, gyro_data.z};
+
+						// Time as ullong (nanoseconds)
+						imu_time = static_cast<ullong>(ts * 1000000);
+
+						// Time as time_point
+						using time_point = std::chrono::system_clock::time_point;
+						time_point uptime_timepoint{std::chrono::duration_cast<time_point::duration>(std::chrono::nanoseconds(imu_time))};
+						std::time_t time2 = std::chrono::system_clock::to_time_t(uptime_timepoint);
+						t = std::chrono::system_clock::from_time_t(time2);
+
+						// Images
+						std::optional<cv::Mat*> img0 = std::nullopt;
+						std::optional<cv::Mat*> img1 = std::nullopt;
+						std::optional<cv::Mat*> rgb = std::nullopt;
+						std::optional<cv::Mat*> depth = std::nullopt;
+						if (last_iteration_cam != cam_type_.iteration) {
+							last_iteration_cam = cam_type_.iteration;
+							img0 = cam_type_.img0;
+							img1 = cam_type_.img1;
+							rgb = cam_type_.rgb;
+							depth = cam_type_.depth;
+						}
+
+						// Submit to switchboard
+						_m_imu_cam->put(new imu_cam_type {
+								t,
+								av,
+								la,
+								img0,
+								img1,
+								imu_time,
+						});
+
+						if (rgb && depth) {
+							_m_rgb_depth->put(new rgb_depth_type{
+								rgb,
+								depth
+							});
+						}
+					}
+				}
+    };
+
+		cfg.disable_all_streams();
+		cfg.enable_stream(RS2_STREAM_ACCEL, RS2_FORMAT_MOTION_XYZ32F, ACCEL_RATE); // adjustable to 0, 63 (default), 250 hz
+		cfg.enable_stream(RS2_STREAM_GYRO, RS2_FORMAT_MOTION_XYZ32F, GYRO_RATE); // adjustable set to 0, 200 (default), 400 hz
+		cfg.enable_stream(RS2_STREAM_INFRARED, 1, IMAGE_WIDTH, IMAGE_HEIGHT, RS2_FORMAT_Y8, FPS);
+		cfg.enable_stream(RS2_STREAM_INFRARED, 2, IMAGE_WIDTH, IMAGE_HEIGHT, RS2_FORMAT_Y8, FPS);
+		cfg.enable_stream(RS2_STREAM_COLOR, IMAGE_WIDTH, IMAGE_HEIGHT, RS2_FORMAT_BGR8, FPS);
+		cfg.enable_stream(RS2_STREAM_DEPTH, IMAGE_WIDTH, IMAGE_HEIGHT, RS2_FORMAT_Z16, FPS);
+		profiles = pipe.start(cfg, callback);
+
+		// Get depth scale which is used to convert the measurements into millimeters
+    depth_scale = pipe.get_active_profile().get_device().first<rs2::depth_sensor>().get_depth_scale();
+	}
+
+	virtual void start() override { plugin::start(); }
+
+	virtual void stop() override { plugin::stop(); }
+
+	virtual ~realsense() override { pipe.stop(); }
+
+private:
+	const std::shared_ptr<switchboard> sb;
+	std::unique_ptr<writer<imu_cam_type>> _m_imu_cam;
+	std::unique_ptr<writer<rgb_depth_type>> _m_rgb_depth;
+
+	std::mutex mutex;
+	rs2::pipeline_profile profiles;
+	rs2::pipeline pipe;
+	rs2::config cfg;
+	float depth_scale;
+	rs2_vector gyro_data;
+	rs2_vector accel_data;
+	double ts;
+
+	Eigen::Vector3f la;
```

Comment:
Declare `la` and `av` as locals on L83. Fewer instance variables => less state => fewer things can go wrong.

```C++
Eigen::Vector3f la = {accel.x, accel.y, accel.z};
```

Reply:
Same for the timestamps.